File: server/services/JobManager.js
import fs from 'fs';
import path from 'path';

/**
 * In-Memory & File-backed Asynchronous Processing Job Queue Manager
 */
class JobManager {
  constructor() {
    this.jobs = new Map();
  }

  createJob(projectId, options = {}) {
    const jobId = `job-${Date.now()}-${Math.round(Math.random() * 1000)}`;
    const job = {
      id: jobId,
      projectId,
      status: 'queued',
      progress: 0,
      stage: 'QUEUED',
      stageLabel: 'Queued for Photogrammetry Processing',
      message: 'Job queued in Aero3D reconstruction pipeline.',
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
      inputType: options.inputType || 'video',
      inputFile: options.inputFile || null,
      reconstructionEngine: options.reconstructionEngine || process.env.PHOTOGRAMMETRY_ENGINE || 'demo',
      qualityMetrics: null,
      error: null
    };

    this.jobs.set(jobId, job);
    return job;
  }

  getJob(jobId) {
    return this.jobs.get(jobId) || null;
  }

  getJobByProject(projectId) {
    for (const job of this.jobs.values()) {
      if (job.projectId === projectId) return job;
    }
    return null;
  }

  updateJob(jobId, updates) {
    const job = this.jobs.get(jobId);
    if (!job) return null;

    Object.assign(job, updates, { updatedAt: new Date().toISOString() });
    this.jobs.set(jobId, job);
    return job;
  }

  failJob(jobId, stage, message, details = {}) {
    return this.updateJob(jobId, {
      status: 'failed',
      stage,
      stageLabel: `Failed during ${stage}`,
      message,
      error: {
        stage,
        message,
        details: details.message || 'Check captured frame overlap, texture contrast, and lighting conditions.'
      }
    });
  }
}

export const jobManager = new JobManager();

File: backend/app/pipeline/job_manager.py
import time
import uuid
from typing import Dict, Any, Optional

class JobManager:
    def __init__(self):
        self.jobs: Dict[str, Dict[str, Any]] = {}

    def create_job(self, project_id: str, input_type: str = "video", engine: str = "demo") -> Dict[str, Any]:
        job_id = f"job-{int(time.time())}-{uuid.uuid4().hex[:6]}"
        job = {
            "id": job_id,
            "projectId": project_id,
            "status": "queued",
            "progress": 0,
            "stage": "QUEUED",
            "stageLabel": "Queued for Reconstruction",
            "message": "Job queued in Aero3D Python FastAPI photogrammetry pipeline.",
            "createdAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "updatedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "inputType": input_type,
            "reconstructionEngine": engine,
            "qualityMetrics": None,
            "error": None
        }
        self.jobs[job_id] = job
        return job

    def get_job(self, job_id: str) -> Optional[Dict[str, Any]]:
        return self.jobs.get(job_id)

    def get_job_by_project(self, project_id: str) -> Optional[Dict[str, Any]]:
        for job in self.jobs.values():
            if job["projectId"] == project_id:
                return job
        return None

    def update_job(self, job_id: str, updates: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        if job_id not in self.jobs:
            return None
        self.jobs[job_id].update(updates)
        self.jobs[job_id]["updatedAt"] = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        return self.jobs[job_id]

    def fail_job(self, job_id: str, stage: str, message: str) -> Optional[Dict[str, Any]]:
        return self.update_job(job_id, {
            "status": "failed",
            "stage": stage,
            "stageLabel": f"Failed during {stage}",
            "message": message,
            "error": {"stage": stage, "message": message}
        })

python_job_manager = JobManager()
